language/move-vm/runtime/src/loader.rs

```diff
@@ -590,7 +590,41 @@ impl Loader {
         data_store: &mut impl DataStore,
         log_context: &impl LogContext,
     ) -> VMResult<()> {
-        self.verify_module(module, data_store, true, log_context)
+        // A standard DFS algorithm that follows the dependency chain from the `source_module_id`
```

Comment:
I don't think this check should be put here, let me take a look

Reply:
Yeah, I would put all of this logic in a new file, and invoke it in `verify_module_dependencies`. SImilar to the style where we have a lot of small separate checks in `verify_module`

Reply:
Will do!

Reply:
After a deeper look at the code I realized that there is a difference between putting a check at the `verify_module_verify_no_missing_dependencies` vs `verify_module_dependencies` or `verify_module`:

1) It happens that `verify_module_verify_no_missing_dependencies` is the entry point for module publishing (and it is only called by the VM for module publishing), hence, having the cyclic dependencies check there only takes info effect when a new module is published.

2) However, putting the cyclic dependencies check in `verify_module_dependencies` or `verify_module` means that whenever the module is first loaded into the code cache, the cyclic dependencies check will be performed. And in the case of `A deps on B deps on C deps on D`, the check will run on D, then C -> D, then B -> C -> D, and then A -> B -> C -> D.

I am not so sure which way is better. To me option 2 feels safer although it means a bit longer time for code cache loading.

Also, I would prefer to invoke the check in `verify_module` instead of `verify_module_dependencies` as the latter only checks for *immediate* dependencies of a module while what we need to look at is all dependencies.

Reply:
I wrote this up already somewhere but it got squashed. So sorry if you get this twice. 

This code needs to be changed to be brought inline with the style of all of the other various checkers. Which is (2) but with some other changes
1. Make a new file in bytcode-verifier/src. Name it something like cyclic_dependencies.rs
  * alternatively you could add it as a pass in `DependencyChecker::verify`
2. Move all the code into that new module. The newer checkers that I wrote (or rewrote) have all of the entry points as a `verify` function in the module, e.g. `type_safety::verify`. Other older checks have it as an entry point in a type, e.g. `DependencyChecker::verify`
3. Invoke the `verify` function after `DependencyChecker::verify` in `verify_module`

Reply:
Thank you Todd, I thought about what you suggested before and one thing that hinders me from taking that approach is the need for the `loader`, or to be more precisely, a map from `module_id` to `module`, and the `module_dependencies` function. I also tried to piggyback on `DependencyChecker` but later realized that it only captures the immediate dependencies.

Of course these are not unsolvable problems so I am happy to refactor it to make it inline with the style of other checkers :)

Reply:
I see what you are saying now about the loader. I'd still like to try to refactor it if possible. 

Perhaps we can just have the function take something like `F: FnMut(ModuleID) -> &CompiledModule` or something similar. And this function can then be populated by the loader. Thoughts?

Reply:
I am not sure about the difficulty of having loader to provide a `F: FnMut(ModuleID) -> &CompiledModule`, but one thing we can do to avoid the problem is to pre-calculate `all_deps: BTreeMap<ModuleId, &CompiledModule>` and pass it to `CyclicDependencyChecker::verify(module, &all_deps)`. Then we can do our DFS there.

Reply:
Done, relocated the check over to `bytecode-verifier`, with the help of `F: Fn(&ModuleID) -> Vec<ModuleId>` that can be supplied by the `loader` (and later, in the `move doctor` case, by sth from the `OnDiskStateView`).